File: src/model_settings.py
"""Execution settings metadata, not model capability qualification authority."""
import json
from pathlib import Path

CATALOG = json.loads((Path(__file__).resolve().parents[1] / "public/model-settings-catalog.json").read_text())


def descriptor(provider, model):
    return next((item for item in CATALOG["models"] if (item["provider"], item["model"]) == (provider, model)), None)


def envelope(provider, model, settings=None, revision="default", scope=None):
    item = descriptor(provider, model)
    if not item:
        return None
    return {"contract_version": CATALOG["version"], "connection_id": item["connection_id"],
            "descriptor_revision": item["descriptor_revision"], "settings_schema_version": item["settings_schema_version"],
            "effective_settings": settings if settings is not None else {key: value["default"] for key, value in item["parameters"].items()},
            "selection_revision": revision, "scope": scope}


def validate(value, provider, model):
    item = descriptor(provider, model)
    if not item or not isinstance(value, dict) or set(value) != set(envelope(provider, model)):
        raise ValueError("runtime_execution_settings_invalid")
    expected = envelope(provider, model)
    for key in ("contract_version", "connection_id", "descriptor_revision", "settings_schema_version"):
        if value[key] != expected[key]:
            raise ValueError("runtime_execution_settings_identity_mismatch")
    settings = value["effective_settings"]
    if not isinstance(settings, dict) or set(settings) != set(item["parameters"]):
        raise ValueError("runtime_execution_settings_invalid")
    for key, spec in item["parameters"].items():
        if settings[key] not in [option["value"] for option in spec["options"]]:
            raise ValueError("runtime_execution_settings_unsupported")
    if not isinstance(value["selection_revision"], str) or not 0 < len(value["selection_revision"]) <= 200:
        raise ValueError("runtime_selection_revision_invalid")
    if value["scope"] is not None and (not isinstance(value["scope"], str) or not 0 < len(value["scope"]) <= 300):
        raise ValueError("runtime_selection_scope_invalid")
    return json.loads(json.dumps(value))


def apply_execution_settings(payload, snapshot):
    raw = snapshot.to_dict() if hasattr(snapshot, "to_dict") else snapshot
    value = validate(raw.get("execution_settings"), raw["provider"], raw["model"])
    result = dict(payload)
    # DeepSeek's qualified adapter has no tunable reasoning parameter.
    if raw["provider"] == "codex":
        result["reasoning_effort"] = value["effective_settings"]["reasoning_effort"]
    return result


def fingerprint_settings(snapshot):
    value = snapshot.get("execution_settings")
    if value is None:
        return None
    return {key: value[key] for key in ("contract_version", "connection_id", "descriptor_revision", "settings_schema_version", "effective_settings")}

File: public/model-settings.js
"use strict";
(function (root, factory) {
  const catalog = root.AriadneModelSettingsCatalog || (typeof module === "object" ? require("./model-settings-catalog.json") : null);
  const api = factory(catalog);
  if (typeof module === "object") module.exports = api;
  root.AriadneModelSettings = api;
}(globalThis, function (catalog) {
  const descriptor = (provider, model) => [...(catalog?.models || []), ...(catalog?.retired_models || [])].find(item => item.provider === provider && item.model === model);
  const currentModel = (provider, model) => catalog?.migrations?.[provider]?.[model] || model;
  function envelope(runtime, settings, revision = "default", scope = null) {
    const item = descriptor(runtime.provider, runtime.model);
    if (!item || runtime.mode !== "model") return null;
    return { contract_version: catalog.version, connection_id: item.connection_id, descriptor_revision: item.descriptor_revision,
      settings_schema_version: item.settings_schema_version, effective_settings: settings ?? Object.fromEntries(Object.entries(item.parameters).map(([key, spec]) => [key, spec.default])),
      selection_revision: revision, scope };
  }
  function validate(value, provider, model) {
    const item = descriptor(provider, model), expected = envelope({ mode: "model", provider, model });
    if (!item || !value || JSON.stringify(Object.keys(value).sort()) !== JSON.stringify(Object.keys(expected).sort())) throw Error("runtime_execution_settings_invalid");
    for (const key of ["contract_version", "connection_id", "descriptor_revision", "settings_schema_version"]) if (value[key] !== expected[key]) throw Error("runtime_execution_settings_identity_mismatch");
    const settings = value.effective_settings;
    if (!settings || typeof settings !== "object" || Array.isArray(settings) || JSON.stringify(Object.keys(settings).sort()) !== JSON.stringify(Object.keys(item.parameters).sort())) throw Error("runtime_execution_settings_invalid");
    for (const [key, spec] of Object.entries(item.parameters)) if (!spec.options.some(option => option.value === settings[key])) throw Error("runtime_execution_settings_unsupported");
    if (typeof value.selection_revision !== "string" || !value.selection_revision.length || value.selection_revision.length > 200
      || (value.scope !== null && (typeof value.scope !== "string" || !value.scope.length || value.scope.length > 300))) throw Error("runtime_selection_revision_invalid");
    return Object.freeze({ ...value, effective_settings: Object.freeze({ ...settings }) });
  }
  function identity(runtime) {
    return JSON.stringify([runtime.mode, runtime.provider, runtime.model, runtime.execution_settings?.connection_id,
      runtime.execution_settings?.descriptor_revision, runtime.execution_settings?.settings_schema_version, runtime.execution_settings?.effective_settings]);
  }
  function label(runtime, compact = false) {
    if (runtime?.mode !== "model") return "Local";
    const item = descriptor(runtime.provider, runtime.model);
    const effort = runtime.execution_settings?.effective_settings?.reasoning_effort;
    const suffix = item?.parameters.reasoning_effort?.options.find(option => option.value === effort)?.label;
    return `${item?.[compact ? "compact_label" : "short_label"] || runtime.model}${compact && suffix ? ` · ${suffix}` : ""}`;
  }
  return Object.freeze({ catalog, descriptor, currentModel, envelope, validate, identity, label });
}));
